### ISSUE TITLE
"AssertionError: Failed to download unicodedata!" error during installation.

### ISSUE THREAD

**Issue by klonos** [facelessuser/sublime-backrefs#3]
I have tried installing on Mac OS 10.12.6 multiple times today and it always fails with the same error (either using cached file or with `--no-cache-dir`):

```
pip2 install backrefs --no-cache-dir
Collecting backrefs
  Downloading backrefs-1.0.1.tar.gz (41kB)
    100% |████████████████████████████████| 51kB 411kB/s 
    Complete output from command python setup.py egg_info:
    Downloading: http://www.unicode.org/Public/5.2.0/ucd/UnicodeData.txt --> /private/var/folders/lq/zbxq8ts94_12q54yk30rhwxm0000gn/T/pip-build-uzzEnL/backrefs/tools/unicodedata/5.2.0/UnicodeData.txt
    Traceback (most recent call last):
      File "/private/var/folders/lq/zbxq8ts94_12q54yk30rhwxm0000gn/T/pip-build-uzzEnL/backrefs/setup.py", line 43, in download_unicodedata
        unidatadownload.download_unicodedata(unicodedata.unidata_version)
      File "/private/var/folders/lq/zbxq8ts94_12q54yk30rhwxm0000gn/T/pip-build-uzzEnL/backrefs/tools/unidatadownload.py", line 54, in download_unicodedata
        response = urlopen(furl)
      File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 154, in urlopen
        return opener.open(url, data, timeout)
      File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 429, in open
        response = self._open(req, data)
      File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 447, in _open
        '_open', req)
      File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
        result = func(*args)
      File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1228, in http_open
        return self.do_open(httplib.HTTPConnection, req)
      File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1198, in do_open
        raise URLError(err)
    URLError: <urlopen error [Errno 60] Operation timed out>
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/lq/zbxq8ts94_12q54yk30rhwxm0000gn/T/pip-build-uzzEnL/backrefs/setup.py", line 77, in <module>
        download_unicodedata()
      File "/private/var/folders/lq/zbxq8ts94_12q54yk30rhwxm0000gn/T/pip-build-uzzEnL/backrefs/setup.py", line 50, in download_unicodedata
        assert not fail, "Failed to download unicodedata!"
    AssertionError: Failed to download unicodedata!
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /private/var/folders/lq/zbxq8ts94_12q54yk30rhwxm0000gn/T/pip-build-uzzEnL/backrefs/
```

Any pointers?

**Comment by klonos** [comment]
...just noticed that http://www.unicode.org/Public/5.2.0/ucd/UnicodeData.txt times out (as does http://www.unicode.org in general). Will try again later and report back.

**Comment by facelessuser** [comment]
So what it's doing is it is trying to download the Unicode information for your current Python build.  In this case, for Python 2.7, Unicode 5.2.0 files. In particular it is failing to download http://www.unicode.org/Public/5.2.0/ucd/UnicodeData.txt.  Maybe it's timing out?  Maybe your network is not allowing it?

**Comment by facelessuser** [comment]
Just to elaborate.  It is trying to use the actual official Unicode files to build up the appropriate property patterns so that they match the same Unicode spec that is used for things like `\w` for `re` etc.

**Comment by klonos** [comment]
Thank you for the very prompt response @facelessuser!! ...it seems that it's not the first time that people are having problems conecting to unicode.org: http://help.appveyor.com/discussions/problems/6012-test-fails-to-connect-to-unicodeorg

Was looking for a CDN or something, but the only thing that seems to be working is [ftp://ftp.unicode.org/Public/UNIDATA/UnicodeData.txt](ftp://ftp.unicode.org/Public/UNIDATA/UnicodeData.txt)

**Comment by klonos** [comment]
...or if you want the same, specific version: ftp://ftp.unicode.org/Public/5.2.0/ucd/UnicodeData.txt

**Comment by facelessuser** [comment]
> Thank you for the very prompt response @facelessuser!!

Just screwing around watching youtube videos 🙂 .

> it seems that it's not the first time that people are having problems conecting to unicode.org

Is it a regional issue or something?  I'm connecting to it just fine, but that doesn't mean it isn't having issues.

> or if you want the same, specific version: ftp://ftp.unicode.org/Public/5.2.0/ucd/UnicodeData.txt

It has to be the specific version.  If I was replacing all unicode in re etc. then I could use whatever, but I am only providing Unicode properties, so they must match what is used by that Python build.

Is the ftp vs http supposed to be more reliable?  I guess I could use it as a fallback maybe.  Like maybe if http times out, try the ftp.

**Comment by klonos** [comment]
Is there any way to "manually" download UnicodeData.txt and place it where it expected to be found?

**Comment by klonos** [comment]
> Is it a regional issue or something?

I am down-under :) ...will try tethering my mobile just in case it is ISP-specific.

**Comment by klonos** [comment]
> Is the ftp vs http supposed to be more reliable? I guess I could use it as a fallback maybe. Like maybe if http times out, try the ftp.

Sounds like a plan to me 😉

**Comment by facelessuser** [comment]
I think so...you'd have to download the zip/tar source and place it where it expects

![screenshot 2017-08-05 20 05 28](https://user-images.githubusercontent.com/1055125/28999980-8f1670a4-7a19-11e7-8095-d71328f80d8d.png)

If I recall correctly, it will use it if it is already there...I guess I need to look at the generation script and confirm.

**Comment by facelessuser** [comment]
Yeah, looks like if I find the file, I skip downloading: https://github.com/facelessuser/backrefs/blob/master/tools/unidatadownload.py#L59.

I just realized this issue was created under the Sublime dependency `sublime-backrefs`  this issue should really be under the PyPi package https://github.com/facelessuser/backrefs.  Sublime dependency comes with unicodedata already generated for Sublime's Python3.3.

**Comment by klonos** [comment]
It must be regional, but blocked on the unicode.org side. I was trying with Telstra cable and switching to Vodafone AU tethered from my phone did not help 😞

**Comment by facelessuser** [comment]
But you can confirm that ftp works?

**Comment by klonos** [comment]
>I just realized this issue was creating under the Sublime dependency sublime-backrefs this issue should really be under the PyPi package https://github.com/facelessuser/backrefs. Sublime dependency comes with unicodedata already generated for Sublime's Python3.3.

OOPS! Sorry...

**Comment by klonos** [comment]
>But you can confirm that ftp works?

Yes, ftp works for me.

**Comment by facelessuser** [comment]
You can install manually from source and change the link here to be ftp: https://github.com/facelessuser/backrefs/blob/master/tools/unidatadownload.py#L44.  And then just install from source.

**Comment by klonos** [comment]
![screen shot 2017-08-06 at 12 14 25 pm](https://user-images.githubusercontent.com/2423362/29000010-e4f8a776-7aa0-11e7-9f2f-d783b71d7616.png)

**Comment by klonos** [comment]
>You can install manually form source and change the link here to be ftp: https://github.com/facelessuser/backrefs/blob/master/tools/unidatadownload.py#L44. And then just install from source.

Will give it a go, but totally python newbie 😅

**Comment by klonos** [comment]
...I was actually installing this via pip2 because it is a dependency for https://github.com/facelessuser/Rummage

**Comment by facelessuser** [comment]
Ahh, yeah. that makes sense.

I guess install it locally `python setup.py install` for the time being.

I'll have to cobble together a solution sometime soon.  I just didn't realize this was an issue.

**Comment by klonos** [comment]
Closing in favor of https://github.com/facelessuser/backrefs/issues/5

**Comment by klonos** [comment]
...thank you for all your help @facelessuser. Lets take the discussion to the proper issue queue, shall we?

**Comment by facelessuser** [comment]
👍
